carsus/io/cmfgen/base.py

```diff
@@ -809,3 +836,100 @@ def _get_levels_lines(self, data):
+
+        Parameters
+        ----------
+        data : dict
+           Dictionary containing one dictionary per species with 
+           keys `levels` and `lines`. 
+        temperature_grid : list of numbers, optional
+            Temperatures to have in the collision dataframe. The collision dataframe
+            will have all the temperatures from the CMFGEN dataset by default.
+        """
+        col_list = []
+
+        for ion, data_dict in data.items():
+            levels = data_dict["levels"].copy()
+            collisions = data_dict["collisions"].copy()
+
+            # mapping of label names and their IDs
+            mapping = {
+                label: [id_, g] for label, id_, g in zip(levels.label, levels.ID, levels.g)
+            }
+
+            gi, lower_level_index, upper_level_index = [], [], []
+
+            for ll, ul in zip(collisions.label_lower, collisions.label_upper):
+                if ll in mapping:
+                    lower_level_index.append(mapping[ll][0])
+                    gi.append(mapping[ll][1])
+                else:
+                    lower_level_index.append(np.nan)
+                    gi.append(np.nan)
+                
+                if ul in mapping:
+                    upper_level_index.append(mapping[ul][0])
+                else:
+                    upper_level_index.append(np.nan)
+
+            (
+                collisions["level_number_lower"],
+                collisions["level_number_upper"],
+                collisions["gi"]
+            ) = (lower_level_index, upper_level_index, gi)
+
+            collisions = collisions.dropna(subset=['level_number_lower', 'level_number_upper'])
+            
+            collisions["atomic_number"] = ion[0]
+            collisions["ion_number"] = ion[1]
+            collisions = collisions.drop(columns=["label_lower", "label_upper"])
+
+            collisions = collisions.astype({
+                'level_number_upper': int,
+                'level_number_lower': int,
+            })
+
+            collisions = collisions.set_index(
+                [
+                    "atomic_number",
+                    "ion_number",
+                    "level_number_lower",
+                    "level_number_upper",
+                ]
+            )
+            # divide the dataframe by gi and remove the column
+            collisions = collisions.iloc[:,:-1].div(collisions.gi, axis=0)
+            col_list.append(collisions)
+
+        collisions = pd.concat(col_list)
+
+        collisions.columns = collisions.columns.map(float)
+
+        # assign new columns from the temperature grid
+        if temperature_grid:
+            for item in temperature_grid:
+                if item not in collisions.columns:
+                    collisions.loc[:, item] = np.nan
+
+        collisions = collisions[sorted(collisions.columns)]
+        collisions = collisions.interpolate(axis="columns", limit_direction="both")
```

Comment:
I don't think this works. The DataFrame interpolate method treats the values as equally spaced, which is not necessarily the case. We need an interpolation method that also uses the temperatures as an input such as https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp1d.html .